### PR TITLE
Refactor user and token implementations to be immutable and allow for ad-hoc additional information.

### PR DESCRIPTION
**Pull request by jeff-blaisdell**
Summary:
Attempted to make the model more extensible while more closely matching the Spring implementation we're wrapping:
See [DefaultOAuth2AccessToken](http://docs.spring.io/spring-security/oauth/apidocs/org/springframework/security/oauth2/common/DefaultOAuth2AccessToken.html) for design inspiration.

Key changes:
- Refactored to allow for custom ad-hoc properties to be exposed via an additionalInformation field on both the OAuthToken and User implementations.
- Made both User and OAuthToken models immutable.
- Renamed `scopes` field to match Spring's `scope'`
- Renamed access on OAuthToken from `getAccessToken` to match Spring's `getValue`
- ObjectMapper no longer exposed for configuration.
- Bumped version to 2.0

### CODE REVIEW

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckTokenValidator.java [hunk 1]
```diff
@@ -93,4 +107,16 @@ private String buildBasicAuthHeader(String user, String password) {
...
+		ObjectMapper objectMapper = new ObjectMapper();
+		objectMapper.registerModule(new Jdk7Module());
+		objectMapper.registerModule(new Jdk8Module());
+		objectMapper.registerModule(new GuavaModule());
+		objectMapper.setPropertyNamingStrategy(PropertyNamingStrategy.CAMEL_CASE_TO_LOWER_CASE_WITH_UNDERSCORES);
+		objectMapper.getFactory().enable(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES);
+		objectMapper.getFactory().enable(JsonParser.Feature.ALLOW_SINGLE_QUOTES);
+		objectMapper.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+		return objectMapper;
+	}
```
Comment (jeff-blaisdell) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckTokenValidator.java:101]: This makes the ObjectMapper no longer configurable via Guice injection.  

I did this to address the specific nature of the JSON being served up by spring w/ snake cased properties.
Other possible options:
a) leave it configurable and utilize `@JsonProperty` annotations.
b) Don't use `@JsonProperty` and instead require ObjectMapper to be configured as expected.

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckAuthModule.java [hunk 2]
```diff
@@ -18,14 +16,16 @@ protected void configure() {
...
-		return new CachingTokenValidator(new SpringSecCheckTokenValidator(config, httpClient, objectMapper));
+	public TokenValidator tokenValidator(SpringSecCheckAuthModule.Config config, HttpClient httpClient) {
+		return new CachingTokenValidator(new SpringSecCheckTokenValidator(config, httpClient));
 	}
 
 	public static class Config {
 		private URI host;
 		private String user;
 		private String password;
+		private String userImpl = DefaultUser.class.getCanonicalName();
```
Comment (llinder) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckAuthModule.java]: Is it worth thinking about making this Class<? extends User>?

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultUser.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package st.ratpack.auth;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Set;
+
+public class DefaultUser implements User {
```
Comment (llinder) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultUser.java:5]: Any objections to making user implementations immutable?

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultOAuthToken.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+package st.ratpack.auth;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Optional;
+import java.util.Set;
+
+public class DefaultOAuthToken implements OAuthToken {
```
Comment (llinder) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultOAuthToken.java:5]: Token implementations really feel like something that should be immutable.

**Comment by llinder** [comment]
Using interfaces for the access token and user is a good change.  My personal view is that the interface and default implementations should enforce immutability though.

I have to wonder if exposing token, user and object mapper as configurable values is entirely necessary though?  For example, Spring Security has the notion of extra attributes on the access token.  One could easily create a new type that wraps the underlying access token and exposes the map of extra stuff as typed properties.

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultUser.java (hunk 3 above)
Comment (beckje01) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/DefaultUser.java:5]: That should be fine.

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckAuthModule.java (hunk 2 above)
Comment (beckje01) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckAuthModule.java]: Sounds a bit better....

File: ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckTokenValidator.java (hunk 1 above)
Comment (beckje01) [inline ratpack-bearer-auth/src/main/java/st/ratpack/auth/springsec/SpringSecCheckTokenValidator.java:101]: Yeah this seems fine to me.

**Comment by jeff-blaisdell** [comment]
I'm fine w/ that.  It seems like a good direction and it solves the same problem I'm facing.

That would make this a backwards incompatible change.  If we took it that far, how do you feel about dropping the User model entirely.  I'm not really seeing the point as the OAuthToken model would have everything anyways.  It also seems to better match Spring Securities impl as well, [DefaultOAuth2AccessToken](http://docs.spring.io/spring-security/oauth/apidocs/org/springframework/security/oauth2/common/DefaultOAuth2AccessToken.html)

**Comment by jeff-blaisdell** [comment]
Incorporated feedback and finished a pass at that direction.

**Comment by llinder** [comment]
Looks like a good direction to me 👍

**Pull request merged**